### PR TITLE
Turn some calls to `underlying` into `superType`.

### PR DESCRIPTION
**Pull request by odersky**
Based on #15453

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -4157,29 +4164,29 @@ object Types {
...
-    private var cachedSuper: Type = _
+    private var cachedSuper: Type = uninitialized
 
     // Boolean caches: 0 = uninitialized, -1 = false, 1 = true
     private var myStableHash: Byte = 0
     private var myGround: Byte = 0
 
-    private var myIsStablePeriod: Period = Nowhere
-    private var myIsStable: Boolean = false
+    private var myisStableRunId: RunId = NoRunId
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/core/Types.scala:4173]:
```suggestion
    private var myIsStableRunId: RunId = NoRunId
```

File: compiler/test-resources/repl/i6474 [hunk 2]
```diff
@@ -5,20 +5,20 @@ scala> object Foo2 { type T[+A] = [B] =>> (A, B) }
 scala> object Foo3 { type T[+A] = [B] =>> [C] =>> (A, B) }
 // defined object Foo3
 scala> ((1, 2): Foo1.T[Int]): Foo1.T[Any]
-val res0: (Any, Int) = (1,2)
+val res0: Foo1.T[Any] = (1,2)
```
Comment (dwijnand) [inline compiler/test-resources/repl/i6474:8]: That's nice

**Comment by nicolasstucki** [comment]
To fix the MiMa tests, we should be able to re-run the tests after  #15457 is merged (no-rebase needed).

**Comment by odersky** [comment]
test performance  please

**Comment by odersky** [comment]
test performance please

**Review by mbovel — APPROVED** [review]

**Pull request merged**